src/mscorlib/src/System/Threading/ExecutionContext.cs

```diff
@@ -191,18 +191,27 @@ internal static void SetLocalValue(IAsyncLocal local, object newValue, bool need
             if (previousValue == newValue)
                 return;
 
-            current = new ExecutionContext(current);
-            current.m_localValues[local] = newValue;
-
-            t_currentMaybeNull = current;
+            Dictionary<IAsyncLocal, object> newValues = new Dictionary<IAsyncLocal, object>(current.m_localValues);
+            newValues[local] = newValue;
 
+            List<IAsyncLocal> newChangeNotifications = current.m_localChangeNotifications;
             if (needChangeNotifications)
             {
                 if (hadPreviousValue)
-                    Contract.Assert(current.m_localChangeNotifications.Contains(local));
+                {
+                    Contract.Assert(newChangeNotifications.Contains(local));
+                }
                 else
-                    current.m_localChangeNotifications.Add(local);
+                {
+                    newChangeNotifications = new List<IAsyncLocal>(current.m_localChangeNotifications);
+                    newChangeNotifications.Add(local);
+                }
+            }
+
+            t_currentMaybeNull = new ExecutionContext(newValues, newChangeNotifications);
```

Comment:
Yes.  We can probably also avoid allocating a new list/dictionary in the cases where the ExecutionContext hasn't been captured.  However, I'm skeptical that this will have much benefit in practice, as ExecutionContext is captured a *lot*.  For example, every Async method captures the current ExecutionContext when it's invoked, even if it never awaits anything.